sdk/include/platform/sunburst/platform-ethernet.hh

```diff
@@ -0,0 +1,633 @@
+		return (status & Port1Status::LinkGood) != 0;
+	}
+
+	std::optional<Frame> receive_frame()
+	{
+		if (frames_to_process == 0)
+		{
+			uint16_t isr = register_read(RegisterOffset::ISR);
+			if (!(isr & ReceiveInterrupt))
+			{
+				return std::nullopt;
+			}
+
+			// Acknowledge the interrupt
+			register_write(RegisterOffset::ISR, ReceiveInterrupt);
+
+			// Read number of frames pending.
+			// Note that this is only updated when we acknowledge the interrupt.
+			frames_to_process = register_read(RegisterOffset::RXFCTR) >> 8;
+		}
+
+		// Get number of frames pending
+		for (; frames_to_process; frames_to_process--)
+		{
+			uint16_t status = register_read(RegisterOffset::RXFHSR);
+			uint16_t length = register_read(RegisterOffset::RXFHBCR) & 0xFFF;
+			bool     valid =
+			  (status & RxFrameValid) &&
+			  !(status &
+			    (RxCrcError | RxRuntFrame | RxFrameTooLong | RxMiiError |
+			     RxUdpFrameChecksumStatus | RxTcpFrameChecksumStatus |
+			     RxIpFrameChecksumStatus | RxIcmpFrameChecksumStatus));
+
+			if (!valid && DebugDroppedFrames)
+			{
+				if (DebugDroppedFrames)
+				{
+					Debug::log("Dropping frame with status: {}", status);
+				}
+
+				drop_error_frame();
+				continue;
+			}
+
+			if (length == 0)
+			{
+				if (DebugDroppedFrames)
+				{
+					Debug::log("Dropping frame with zero length");
+				}
+
+				drop_error_frame();
+				continue;
+			}
+
+			Debug::log("Receiving frame of length {}", length);
+
+			auto frame = Frame(length);
+
+			// Reset QMU RXQ frame pointer to zero and start DMA transfer
+			// operation.
+			register_write(RegisterOffset::RXFDPR, 0x4000);
+			register_set(RegisterOffset::RXQCR, StartDmaAccess);
+
+			// Start receiving via SPI.
+			uint8_t cmd = 0b10 << 6;
+			set_gpio_output_bit(EthCsPin, 0);
+			spi()->blocking_write(&cmd, 1);
+
+			// Initial words are RXFHSR and RXFHBCR which we have already know
+			// the value.
+			uint8_t dummy[8];
+			spi()->blocking_read(dummy, 8);
+
+			spi()->blocking_read(frame.buffer, length);
+
+			// Receive needs to be dword-aligned, read the paddings.
+			uint16_t pad = (-length) & 0x3;
+			if (pad != 0)
+			{
+				spi()->blocking_read(dummy, pad);
+			}
+
+			set_gpio_output_bit(EthCsPin, 1);
+
+			register_clear(RegisterOffset::RXQCR, StartDmaAccess);
+			frames_to_process -= 1;
+
+			return std::move(frame);
+		}
+
+		return std::nullopt;
+	}
+
+	/**
+	 * Send a packet.  This will block if no buffer space is available on
+	 * device.
+	 *
+	 * The third argument is a callback that allows the caller to check the
+	 * frame before it's sent but after it's copied into memory that isn't
+	 * shared with other compartments.
+	 */
+	bool send_frame(const uint8_t *buffer, uint16_t length, auto &&check)
+	{
+		// We must check the frame pointer and its length. Although it
+		// is supplied by the firewall which is trusted, the firewall
+		// does not check the pointer which is coming from external
+		// untrusted components.
```

Comment:
The firewall does not check because the threat model assumes that a DoS on the firewall by the T P/IP stack does not make sense. The hardware-interface layer of a driver should not be defining a threat model.

Reply:
This part is copied from the Arty A7 driver.

I have to call the `check` function from within this function. Are you suggesting that I call `check(buffer)` directly? Is there a guarantee that the buffer won't change during the execution of `send_frame`? Otherwise you'll get TOCTOU issue.